lib/ckb/udt_wallet.rb

```diff
@@ -238,17 +290,111 @@ def gather_inputs(amount)
       OpenStruct.new(inputs: inputs, amounts: input_amounts,
                      capacities: input_capacities)
     end
+  end
 
-    def pubkey_bin
-      Ckb::Utils.extract_pubkey_bin(privkey)
+  class UdtCellWallet < UdtBaseWallet
+    def get_balance
+      fetch_cell[:amount]
+    end
+
+    def latest_outpoint
+      fetch_cell[:outpoint]
+    end
+
+    def created?
+      get_unspent_cells.length > 0
+    end
+
+    def fetch_cell
+      cells = get_unspent_cells
+      case cells.length
+      when 0
+        raise "Please create udt cell wallet first!"
+      when 1
+        cells[0]
+      else
+        raise "There's more than one cell for this UDT! You can use merge_cells in UdtWallet to merge them into one"
+      end
     end
 
-    def self.random(api)
-      self.new(api, SecureRandom.bytes(32))
+    # Generates a partial tx that provides some UDTs for other user, who
+    # can only accept the exact amount provided here but no more
+    def send_tokens(amount, target_wallet)
+      cell = fetch_cell
+      target_cell = target_wallet.fetch_cell
+      if amount > cell[:amount]
+        raise "Do not have that much amount!"
+      end
+      inputs = [
+        {
+          previous_output: {
+            hash: cell[:outpoint][:hash],
+            index: cell[:outpoint][:index]
+          },
+          unlock: mruby_unlock_script_json_object
+        }
+      ]
+      outputs = [
+        {
+          capacity: cell[:capacity],
+          data: [cell[:amount] - amount].pack("Q<"),
+          lock: address,
+          contract: {
+            version: 0,
+            args: [],
+            reference: api.mruby_cell_hash,
+            signed_args: [
+              Ckb::CONTRACT_SCRIPT,
+              token_info.name,
+              token_info.pubkey
+            ]
+          }
+        },
+        {
+          capacity: target_cell[:capacity],
+          data: [target_cell[:amount] + amount].pack("Q<"),
+          lock: target_cell[:lock],
+          contract: {
+            version: 0,
+            args: [],
+            reference: api.mruby_cell_hash,
+            signed_args: [
+              Ckb::CONTRACT_SCRIPT,
+              token_info.name,
+              token_info.pubkey
+            ]
+          }
+        }
+      ]
```

Comment:
I feel it would be better to check the sum of two outputs equals the original cell[:amount], to prevent integer overflow on line #355 in case.

Reply:
I totally agree we should prevent integer overflow, but I believe we should do the check at [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/cell-wallet/contracts/udt/contract.rb#L56-L70) in the validator. The commented code here is the generator part which should be relatively safe.

Thoughts?